src/tslint/unifiedSignaturesRule.ts

```diff
@@ -24,241 +46,284 @@ export class Rule extends Lint.Rules.AbstractRule {
 			}
 		}
 	}
-
-	private fail(node: ts.Node, message: string) {
-		this.addFailure(this.createFailure(node.getStart(), node.getWidth(), message));
-	}
-}
-
-function signatureName(node: ts.SignatureDeclaration): string | undefined {
-	if (node.kind === ts.SyntaxKind.CallSignature) {
-		return "()";
-	}
-	else {
-		return node.name && getTextOfPropertyName(node.name);
-	}
 }
 
-function collectOverloads(signatures: ts.SignatureDeclaration[]): Map<string, ts.SignatureDeclaration[]> {
-	const map = new Map<string, ts.SignatureDeclaration[]>();
-	for (const sig of signatures) {
-		const name = signatureName(sig);
-		if (name !== undefined) {
-			const got = map.get(name);
-			if (got) {
-				got.push(sig);
-			}
-			else {
-				map.set(name, [sig]);
-			}
-		}
-	}
-	return map;
+function signaturesCanBeUnified(a: ts.SignatureDeclaration, b: ts.SignatureDeclaration, isTypeParameter: IsTypeParameter): boolean {
+	// Must return the same type.
+	return typesAreEqual(a.type, b.type) &&
+		// Must take the same type parameters.
+		arraysAreEqual(a.typeParameters, b.typeParameters, typeParametersAreEqual) &&
+		// If one uses a type parameter (from outside) and the other doesn't, they shouldn't be joined.
+		signatureUsesTypeParameter(a, isTypeParameter) === signatureUsesTypeParameter(b, isTypeParameter);
 }
 
 /** Detect `a(x: number, y: number, z: number)` and `a(x: number, y: string, z: number)`. */
-function signaturesDifferBySingleParameter(types1: string[], types2: string[]): [string, string] | undefined {
-	if (types1.length !== types2.length) {
-		return undefined;
-	}
-
-	const index = getIndexOfFirstDifference(types1, types2);
+function signaturesDifferBySingleParameter(types1: ts.ParameterDeclaration[], types2: ts.ParameterDeclaration[],
+	): [ts.ParameterDeclaration, ts.ParameterDeclaration] | undefined {
+	const index = getIndexOfFirstDifference(types1, types2, parametersAreEqual);
 	if (index === undefined) {
 		return undefined;
 	}
 
 	// If remaining arrays are equal, the signatures differ by just one parameter type
-	if (!arraysEqual(types1.slice(index + 1), types2.slice(index + 1))) {
+	if (!arraysAreEqual(types1.slice(index + 1), types2.slice(index + 1), parametersAreEqual)) {
 		return undefined;
 	}
 
 	const a = types1[index];
 	const b = types2[index];
-	// Must have equivalent optionality / rest-ness
-	if (a.startsWith("?") !== b.startsWith("?") || a.startsWith("...") !== b.startsWith("...")) {
+	return parametersHaveEqualSigils(a, b) ? [a, b] : undefined;
+}
+
+/**
+ * Detect `a(): void` and `a(x: number): void`.
+ * Returns the parameter declaration (`x: number` in this example) that should be optional.
+ */
+function signaturesDifferByOptionalParameter(types1: ts.ParameterDeclaration[], types2: ts.ParameterDeclaration[],
+	): ts.ParameterDeclaration | undefined {
+	const minLength = Math.min(types1.length, types2.length);
+	// Lengths must differ by 1.
+	if (Math.abs(types1.length - types2.length) !== 1) {
```

Comment:
You're right. We should catch `f(): void` and `f(x?: number, y?: number): void`. Will fix later.